In [2]:
import torch
import gc
torch.cuda.empty_cache()
gc.collect()


22

In [3]:
from setup import setup_src_path
print(setup_src_path())
import data.processed as processed
import config.config as config
import utils.setup as setup
import utils.functions as fn
from importlib import reload



['/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/mlm', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '', '/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages', '/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/modules', '/tmp/tmp0tyt5zy3']


In [5]:
from datasets import load_from_disk,concatenate_datasets


adapter_name="mlm_inv_TR"
dataset = load_from_disk(f"{config.Config.DATASETS_SAVE_PATH}/datasets")



In [6]:
reload(config)

    
filtered_target = dataset['train'].filter(lambda example: example['genre'] == "travel").shuffle(seed=42)
#filtered_target = shuffled_filtered_target.train_test_split(test_size=0.1)


filtered_test_target = dataset['validation_matched'].filter(lambda example: example['genre'] == "travel")
train_target = filtered_target
test_target = filtered_test_target
train_target[1]
 

{'promptID': 82447,
 'pairID': '82447e',
 'premise': 'Wherever the original furnishings and decoration were missing, superb appropriate equivalents have been installed.',
 'premise_binary_parse': '( ( Wherever ( ( the ( original ( ( furnishings and ) decoration ) ) ) ( were missing ) ) ) ( , ( ( ( superb appropriate ) equivalents ) ( ( have ( been installed ) ) . ) ) ) )',
 'premise_parse': '(ROOT (S (S (VP (VB Wherever) (SBAR (S (NP (DT the) (JJ original) (NNS furnishings) (CC and) (NN decoration)) (VP (VBD were) (ADJP (VBG missing))))))) (, ,) (NP (ADJP (JJ superb) (JJ appropriate)) (NNS equivalents)) (VP (VBP have) (VP (VBN been) (VP (VBN installed)))) (. .)))',
 'hypothesis': 'Superb equivalents have been installed in the place of the missing original furnishings.',
 'hypothesis_binary_parse': '( ( Superb equivalents ) ( ( have ( been ( installed ( in ( ( the place ) ( of ( the ( missing ( original furnishings ) ) ) ) ) ) ) ) ) . ) )',
 'hypothesis_parse': '(ROOT (S (NP (NNP Superb

In [7]:
from adapters import AutoAdapterModel,init
from tqdm import tqdm

from transformers import AutoConfig, AutoModelForMaskedLM

mdlcfg = AutoConfig.from_pretrained(
    config.Config.MODEL_NAME,
 
)
model = AutoModelForMaskedLM.from_pretrained(
    config.Config.MODEL_NAME,
)
init(model)
reload(fn)
fn.print_trainable_parameters(model)

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertFor

trainable params: 109514298 || all params: 109514298 || trainable%: 100.0


In [8]:
from adapters import ConfigUnion, LoRAConfig, PrefixTuningConfig, SeqBnConfig,SeqBnInvConfig,AdapterConfig,LoRAConfig


adapter_config = SeqBnInvConfig(reduction_factor=2)

model.add_adapter(adapter_name, config=adapter_config)

model.train_adapter([adapter_name])
model.active_adapters = adapter_name
print(model.adapter_summary())
fn.print_trainable_parameters(model)

Name                     Architecture         #Param      %Param  Active   Train
--------------------------------------------------------------------------------
mlm_inv_TR               bottleneck        7,387,776       6.785       1       1
--------------------------------------------------------------------------------
Full model                               108,891,648     100.000               0
trainable params: 7979904 || all params: 116902074 || trainable%: 6.82614407679371


In [9]:
from transformers import AutoTokenizer

reload(processed)
tokenizer = AutoTokenizer.from_pretrained(config.Config.TOKENIZER_NAME)

tokenized_dataset= processed.tokenize_dataset(train_target,tokenizer)
tokenized_dataset

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids'],
    num_rows: 77350
})

In [10]:
chunk_size = 128
# Slicing produces a list of lists for each feature
tokenized_samples = tokenized_dataset[444:470]

results = fn.group_texts(tokenized_samples, chunk_size)
for chunk in results["labels"]:
    print(f"'>>> Chunk length: {len(chunk)}'")

'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'


In [11]:
# we group texts and chunk them
lm_datasets = tokenized_dataset.map(fn.group_texts, batched=True,fn_kwargs={'chunk_size': chunk_size})
lm_datasets

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
    num_rows: 27252
})

In [12]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
samples = [lm_datasets[i] for i in range(2)]
for sample in samples:
    _ = sample.pop("word_ids")
dd=data_collator(samples)
for chunk in dd["input_ids"]:
    print(f"\n'>>> {tokenizer.decode(chunk)}'")


'>>> [CLS] [MASK] you reach the atlantic coast, you'll become almost engulfed in [MASK] [MASK] of banana trees, [MASK] [MASK]es [MASK] tempting [MASK] close to the [MASK] [MASK] [SEP] vast groves of banana trees will almost engulf you. [SEP] [CLS] wherever the original furnishings [MASK] decoration were missing, superb appropriate equivalents have been [MASK]. [SEP] superb [MASK]s have [MASK] installed in the place of the missing original [MASK] electorate [SEP] [CLS] in a low - rise building [MASK] [MASK] [MASK] entrance is a small ethn [MASK] museum, a place to see displays [MASK] local island customs and costumes as you wait deteriorating [MASK] of minutes for the guided cave tour to begin. [SEP] there is'

'>>> a small museum near the cave entrance. [SEP] [CLS] by allowing the local pride of [MASK] historic regions as provence, [MASK], brittany, and langue? doc to re [MASK]rt itself, france demonstrated that it was at last secure in [MASK] national identity'so secure that [MASK] c

In [13]:

downsampled_dataset = lm_datasets.train_test_split(
    test_size=0.1, seed=42
)
downsampled_dataset

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 24526
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 2726
    })
})

In [14]:

reload(fn)
trainer = fn.train_mlm_model(model,adapter_name,data_collator,tokenizer, downsampled_dataset['train'],downsampled_dataset['test'])

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/accelerate/accelerator.py:446: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(


In [15]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/86 [00:00<?, ?it/s]

>>> Perplexity: 13.33


In [16]:
trainer.train()

  0%|          | 0/15340 [00:00<?, ?it/s]

{'loss': 2.1818, 'grad_norm': 1.7855619192123413, 'learning_rate': 9.820754716981132e-05, 'epoch': 1.0}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.8513665199279785, 'eval_runtime': 4.4313, 'eval_samples_per_second': 615.173, 'eval_steps_per_second': 19.407, 'epoch': 1.0}
{'loss': 1.9399, 'grad_norm': 1.7761797904968262, 'learning_rate': 9.304582210242589e-05, 'epoch': 2.0}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.7524170875549316, 'eval_runtime': 4.5033, 'eval_samples_per_second': 605.328, 'eval_steps_per_second': 19.097, 'epoch': 2.0}
{'loss': 1.8721, 'grad_norm': 1.7715872526168823, 'learning_rate': 8.788409703504043e-05, 'epoch': 3.0}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.7160929441452026, 'eval_runtime': 4.5632, 'eval_samples_per_second': 597.381, 'eval_steps_per_second': 18.846, 'epoch': 3.0}
{'loss': 1.8165, 'grad_norm': 1.6436455249786377, 'learning_rate': 8.272237196765499e-05, 'epoch': 3.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.7126827239990234, 'eval_runtime': 4.676, 'eval_samples_per_second': 582.973, 'eval_steps_per_second': 18.392, 'epoch': 4.0}
{'loss': 1.7833, 'grad_norm': 1.7623934745788574, 'learning_rate': 7.756064690026954e-05, 'epoch': 4.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.6776100397109985, 'eval_runtime': 4.3558, 'eval_samples_per_second': 625.837, 'eval_steps_per_second': 19.744, 'epoch': 5.0}
{'loss': 1.7561, 'grad_norm': 1.7034921646118164, 'learning_rate': 7.239892183288411e-05, 'epoch': 5.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.6726394891738892, 'eval_runtime': 7.9876, 'eval_samples_per_second': 341.28, 'eval_steps_per_second': 10.767, 'epoch': 6.0}
{'loss': 1.735, 'grad_norm': 1.8266328573226929, 'learning_rate': 6.723719676549865e-05, 'epoch': 6.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.6441035270690918, 'eval_runtime': 7.7806, 'eval_samples_per_second': 350.357, 'eval_steps_per_second': 11.053, 'epoch': 7.0}
{'loss': 1.7119, 'grad_norm': 1.7161520719528198, 'learning_rate': 6.20822102425876e-05, 'epoch': 7.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.623020887374878, 'eval_runtime': 7.976, 'eval_samples_per_second': 341.775, 'eval_steps_per_second': 10.782, 'epoch': 8.0}
{'loss': 1.7067, 'grad_norm': 1.671831488609314, 'learning_rate': 5.6927223719676545e-05, 'epoch': 8.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.6228677034378052, 'eval_runtime': 7.9289, 'eval_samples_per_second': 343.806, 'eval_steps_per_second': 10.846, 'epoch': 9.0}
{'loss': 1.6913, 'grad_norm': 1.8414323329925537, 'learning_rate': 5.176549865229111e-05, 'epoch': 9.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.6054744720458984, 'eval_runtime': 8.0043, 'eval_samples_per_second': 340.566, 'eval_steps_per_second': 10.744, 'epoch': 10.0}
{'loss': 1.6799, 'grad_norm': 1.7718124389648438, 'learning_rate': 4.660377358490566e-05, 'epoch': 10.99}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.6049340963363647, 'eval_runtime': 7.9459, 'eval_samples_per_second': 343.072, 'eval_steps_per_second': 10.823, 'epoch': 11.0}
{'loss': 1.6604, 'grad_norm': 1.8321046829223633, 'learning_rate': 4.1442048517520216e-05, 'epoch': 11.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5821516513824463, 'eval_runtime': 7.9376, 'eval_samples_per_second': 343.429, 'eval_steps_per_second': 10.835, 'epoch': 12.0}
{'loss': 1.653, 'grad_norm': 1.8702316284179688, 'learning_rate': 3.628032345013477e-05, 'epoch': 12.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5895596742630005, 'eval_runtime': 7.9175, 'eval_samples_per_second': 344.302, 'eval_steps_per_second': 10.862, 'epoch': 13.0}
{'loss': 1.6493, 'grad_norm': 1.8198981285095215, 'learning_rate': 3.1118598382749325e-05, 'epoch': 13.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5820671319961548, 'eval_runtime': 7.9633, 'eval_samples_per_second': 342.322, 'eval_steps_per_second': 10.8, 'epoch': 14.0}
{'loss': 1.6341, 'grad_norm': 1.7574244737625122, 'learning_rate': 2.5963611859838276e-05, 'epoch': 14.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5458641052246094, 'eval_runtime': 8.1241, 'eval_samples_per_second': 335.545, 'eval_steps_per_second': 10.586, 'epoch': 15.0}
{'loss': 1.6271, 'grad_norm': 1.736909031867981, 'learning_rate': 2.080188679245283e-05, 'epoch': 15.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5766187906265259, 'eval_runtime': 7.9108, 'eval_samples_per_second': 344.591, 'eval_steps_per_second': 10.871, 'epoch': 16.0}
{'loss': 1.6247, 'grad_norm': 1.9784667491912842, 'learning_rate': 1.5640161725067385e-05, 'epoch': 16.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5440337657928467, 'eval_runtime': 8.0533, 'eval_samples_per_second': 338.494, 'eval_steps_per_second': 10.679, 'epoch': 17.0}
{'loss': 1.6136, 'grad_norm': 2.019458293914795, 'learning_rate': 1.0485175202156335e-05, 'epoch': 17.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5442390441894531, 'eval_runtime': 7.642, 'eval_samples_per_second': 356.712, 'eval_steps_per_second': 11.254, 'epoch': 18.0}
{'loss': 1.6164, 'grad_norm': 1.787548542022705, 'learning_rate': 5.3234501347708894e-06, 'epoch': 18.98}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5491780042648315, 'eval_runtime': 7.9619, 'eval_samples_per_second': 342.38, 'eval_steps_per_second': 10.801, 'epoch': 19.0}
{'loss': 1.6077, 'grad_norm': 1.9038749933242798, 'learning_rate': 1.684636118598383e-07, 'epoch': 19.97}


  0%|          | 0/86 [00:00<?, ?it/s]

{'eval_loss': 1.5576114654541016, 'eval_runtime': 8.0937, 'eval_samples_per_second': 336.805, 'eval_steps_per_second': 10.626, 'epoch': 20.0}


Overwriting existing adapter 'mlm_inv_TR'.


{'train_runtime': 2896.669, 'train_samples_per_second': 169.339, 'train_steps_per_second': 5.296, 'train_loss': 1.7277888535520711, 'epoch': 20.0}


TrainOutput(global_step=15340, training_loss=1.7277888535520711, metrics={'train_runtime': 2896.669, 'train_samples_per_second': 169.339, 'train_steps_per_second': 5.296, 'train_loss': 1.7277888535520711, 'epoch': 20.0})

In [17]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/86 [00:00<?, ?it/s]

>>> Perplexity: 4.63


In [18]:
import config.config as config

trainer.model.save_adapter(f"{config.Config.ADAPTER_SAVE_PATH}/{adapter_name}", adapter_name)


: 

#### This is a report of mlm for TRAVEL domain using whole genre as target and then splitt after tokenization with seed 42. I used computer metrix for preplexity and early stopping 
The results before:
>>> Perplexity: 13.33

The results after:
>>> Perplexity: 4.63